### PR TITLE
Develop tmr

### PR DESCRIPTION
**Pull request by shakesy94**
This PR is to merge in the time_matching_requirement changes. To do so I:

1- I created a develop_tmr branch from develop

This was done to ensure that changes do not affect the develop branch. 

2- I created a branch from the original time_matching_requirement branch named time_matching_requirement_3

This is done to keep time_matching_requirement branch as a back-up. 

3- I created a pull request from time_matching_requirement_3 into develop_tmr

4- Resolved conflicts in time_matching_requirement_3 into develop_tmr PR

The PR resulted in multiple types of conflicts. Conflicts in the GenX subdirectory were mostly due to updated that were already implemented in the GenX submodule. There was a specific change required in the ESR requirement that needed to be ported. Because the ESR module was re-written in GenX, we opted to forgo the changes, and then add them as part of the merging develop into develop_tmr PR. Other conflicts were resolved. 

We also deleted the the src/GenX folder completely. 

5- Merged time_matching_requirement_3 into develop_tmr PR

GenX submodule got deleted so had to add it back. To add it back I did the following commands in Command Line. 
- git checkout develop_tmr
- rm -rf src/GenX 
- git submodule add --force https://github.com/macroenergy/GenX.git src/GenX
- git submodule update --init --recursive
- I then committed and pushed the changes in github dekstop

### CODE REVIEW

File: src/HSC/model/transmission/h2_pipeline.jl [hunk 1]
```diff
@@ -187,8 +187,6 @@ function h2_pipeline(EP::Model, inputs::Dict, setup::Dict)
 
     EP[:eH2Balance] += ePipeZoneDemand
 
-    EP[:eHTransmissionByZone] += ePipeZoneDemand
```
Comment (shakesy94) [inline src/HSC/model/transmission/h2_pipeline.jl:190]: Track where this change came from. @dharik13 is this on purpose and correct?

File: src/HSC/write_outputs/write_h2_elec_costs.jl [hunk 2]
```diff
@@ -35,8 +35,11 @@ function write_h2_elec_costs(path::AbstractString, sep::AbstractString, inputs::
         dfPrice = DataFrame(transpose(dual.(EP[:cPowerBalance])./inputs["omega"]), :auto)
     end
 
-    dfP2G = DataFrame()
-    dfElecCost = DataFrame()
+	# Recreate ELEC PRICE Dataframe (like in GenX, write_price function)
+#	dfPrice = DataFrame(transpose(dual.(EP[:cPowerBalance])./inputs["omega"]), :auto)
+
+	dfP2G = DataFrame()
+	dfElecCost = DataFrame()
```
Comment (shakesy94) [inline src/HSC/write_outputs/write_h2_elec_costs.jl:42]: Indentation is wrong

File: src/HSC/write_outputs/write_h2_elec_costs.jl [hunk 3]
```diff
@@ -53,8 +56,8 @@ function write_h2_elec_costs(path::AbstractString, sep::AbstractString, inputs::
         append!(dfP2G, tempP2G)
     end
 
-    # Multiply price by power usage, for each zone and each time step
-    dfElecCost = dfP2G .* dfPrice #.* inputs["omega"]
+	# Multiply price by power usage, for each zone and each time step
+	dfElecCost = dfP2G .* dfPrice .* inputs["omega"]
```
Comment (shakesy94) [inline src/HSC/write_outputs/write_h2_elec_costs.jl:60]: indentation is wrong

File: src/generate_model.jl [hunk 4]
```diff
@@ -247,7 +247,13 @@ function generate_model(setup::Dict,inputs::Dict,OPTIMIZER::MOI.OptimizerWithAtt
         end
 
         EP[:eAdditionalDemandByZone] += EP[:eH2NetpowerConsumptionByAll]
-    end
+    
+		# Modeling Time matching requirement for electricity use for hydrogen production
```
Comment (shakesy94) [inline src/generate_model.jl:251]: indentation looks wrong

File: src/generate_model.jl [hunk 5]
```diff
@@ -264,10 +270,10 @@ function generate_model(setup::Dict,inputs::Dict,OPTIMIZER::MOI.OptimizerWithAtt
     end
 
     # Energy Share Requirement
-    if setup["EnergyShareRequirement"] >= 1
+    if setup["EnergyShareRequirement"] == 1
```
Comment (shakesy94) [inline src/generate_model.jl:273]: are there multiple possible ESR requriments

File: src/HSC/model/core/h2_non_served.jl [hunk 6]
```diff
@@ -116,8 +116,6 @@ function h2_non_served(EP::Model, inputs::Dict, setup::Dict)
     # Add non-served energy/curtailed demand contribution to power balance expression
     EP[:eH2Balance] += eH2BalanceNse
 
-    EP[:eHDemandByZone] -= eH2BalanceNse
```
Comment (shakesy94) [inline src/HSC/model/core/h2_non_served.jl:119]: Is this correct @dharik13 ?

**Review by shakesy94 — COMMENTED** [review]
First review

File: src/HSC/model/transmission/h2_pipeline.jl (hunk 1 above)
Comment (shakesy94) [reply src/HSC/model/transmission/h2_pipeline.jl]: added back

File: src/configure_solver/configure_highs.jl [hunk 7]
```diff
@@ -413,520 +413,269 @@ The HiGHS optimizer instance is configured with the following default parameters
     # [type: bool, advanced: true, range: {false, true}, default: true]
     less_infeasible_DSE_check = true
 
-    # Use LiDSE if LP has right properties
-    # [type: bool, advanced: true, range: {false, true}, default: true]
-    less_infeasible_DSE_choose_row = true
-
 """
```
[the remark this answers is not in the document]
Comment (shakesy94) [reply src/configure_solver/configure_highs.jl]: i checked out this file from main. showuld be good to go now.

File: src/generate_model.jl (hunk 5 above)
Comment (shakesy94) [reply src/generate_model.jl]: fixed

File: src/generate_model.jl [hunk 8]
```diff
@@ -190,6 +190,11 @@ function generate_model(setup::Dict,inputs::Dict,OPTIMIZER::MOI.OptimizerWithAtt
         EP = retrofit(EP, inputs)
     end
 
+    # Energy Share Requirement
```
[the remark this answers is not in the document]
Comment (shakesy94) [reply src/generate_model.jl]: done

File: src/HSC/model/policies/time_matching_requirement.jl [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	)
+
+	# Annual excess electricity supply from contracted electricity resources for H2 production
+	@expression(EP, eExcessAnnualElectricitySupplyTMR[TMR=1:nH2_TMR], sum(eExcessElectricitySupplyTMR[TMR,t]*inputs["omega"][t] for t = 1:T))
+
+	## Energy Share Requirements (minimum energy share from qualifying renewable resources) constraint
+	if setup["TimeMatchingRequirement"] == 1 # hourly with excess sales allowed
+	
+		@constraint(EP, cH2TMR[TMR=1:nH2_TMR, t=1:T], eExcessElectricitySupplyTMR[TMR, t]>=0 )
+		# @constraint(EP, cH2TMR[TMR=1:nH2_TMR, t=1:T],
```
Comment (shakesy94) [inline src/HSC/model/policies/time_matching_requirement.jl:79]: I this constraint meant to be commented out? @dharik13

Comment (shakesy94) [reply src/HSC/model/policies/time_matching_requirement.jl:79]: Yes

File: src/HSC/model/policies/time_matching_requirement.jl [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+A complete copy of the GNU General Public License v2 (GPLv2) is available
+in LICENSE.txt.  Users uncompressing this from an archive may not have
+received this license file.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+@doc raw"""
+	energy_share_requirement(EP::Model, inputs::Dict, setup::Dict)
```
[the remark this answers is not in the document]
Comment (shakesy94) [reply src/HSC/model/policies/time_matching_requirement.jl]: added

**Comment by shakesy94** [comment]
@dharik13 This is ready for your review. I think the documentation may need some edits.

**Comment by dharik13** [comment]
09/18:
 - Dharik to review

File: docs/src/hsc_data_documentation.md [hunk 11]
```diff
@@ -32,13 +32,17 @@ Model settings parameters are specified in a `hsc_Settings.yml` file which shoul
...
 ||1 = unit commitment with integer clustering.|
 ||2 = unit commitment with linearized clustering.|
-
 |**Policy related**||
 |H2CO2Cap | Flag for specifying the type of CO2 emission limit constraint.|
 || 0 = no CO2 emission limit|
 || 1 = mass-based emission limit constraint|
 || 2 = load + rate-based emission limit constraint|
 || 3 = generation + rate-based emission limit constraint|
+|TimeMatchingRequirement | Flag for specifying type of time-matching requirement. |
```
Comment (dharik13) [inline docs/src/hsc_data_documentation.md:41]: Added description of flag TMRSalestoESR

File: docs/src/hsc_data_documentation.md [hunk 12]
```diff
@@ -189,7 +193,10 @@ This file contains cost and performance parameters for various generators and ot
 |Up\_Time| Minimum amount of time a resource has to stay in the committed state.|
 |Down\_Time |Minimum amount of time a resource has to remain in the shutdown state.|
 |Start\_Cost\_per\_tonne\_p\_hr |Cost per tonne/hr of nameplate capacity to start a generator ($/tonne/hr per start). Multiplied by the number of generation units (each with a pre-specified nameplate capacity) that is turned on.|
-
+|**TimeMatchingRequirement > 0**||
+|TMR\_*| Flag to indicate which resources are considered for the Time Matching Requirement constraint.|
```
Comment (dharik13) [inline docs/src/hsc_data_documentation.md:197]: Modified to H2_TMR_ . We need to add this to generators_data.csv also when using TMR

File: src/core/expression_manipulation.jl [hunk 13]
```diff
@@ -60,10 +60,10 @@ end
 function add_similar_to_expression!(expr1::Array{GenericAffExpr{C,T}, dim1}, expr2::Array{GenericAffExpr{C,T}, dim2}) where {C,T,dim1,dim2}
     _add_similar_to_expression!(expr1, expr2)
 end
-
-function add_similar_to_expression!(expr1::Array{GenericAffExpr{C,T}, dim1}, expr2::Array{GenericVariableRef{C}, dim2}) where {C,T,dim1,dim2}
-    _add_similar_to_expression!(expr1, expr2)
-end
+# This function threw an error. LoadError: UndefVarError: GenericVariableRef not defined - not sure why
```
Comment (dharik13) [inline src/core/expression_manipulation.jl:63]: Places "GenericVariableRef" was included was giving errors. I commented it out.

**Review by dharik13 — COMMENTED** [review]
@shakesy94 can you please merge to develop after looking to the lines I commented in expression_manipulation.jl

**Review by dharik13 — APPROVED** [review]

**Comment by bennettm** [comment]
> @bennettm can you please merge to develop after looking to the errors I was getting in commented lines in expression_manipulation.jl - to reproduce the error you will have to uncomment the lines

Hi Dharik, do you want me to merge develop_tmr into develop branch?

**Review by bennettm — APPROVED** [review]
I have made minor changes to docstring. I approve this pull request.

**Comment by RuaridhMacd** [comment]
@dharik13 @shakesy94 

I fixed a bug introduced in `write_h2_elec_costs()`

Do we want this example included in ExampleSystems? If so, can we add some documentation for it?

**Pull request merged**